### PR TITLE
Enhance existing Marshal tests

### PR DESCRIPTION
**Pull request by hughbe**
Separated from https://github.com/dotnet/corefx/pull/31057 to make that PR easier to review

/cc @AaronRobinsonMSFT

### CODE REVIEW

**Comment by karelz** [comment]
@luqunl can you please review? It is 4 weeks old PR.

**Comment by hughbe** [comment]
Expecting this to be merged as a single PR, as this is additions to tests rather than cleanup etc. Can’t edit now as I’m without a laptop/git client for the next 3-4 weeks

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/StructureToPtrTests.cs [hunk 1]
```diff
@@ -96,61 +47,65 @@ public void VerifyByValArrayInStruct()
...
             try
             {
-                Marshal.StructureToPtr(structure1, memory, false);
-                Marshal.StructureToPtr(structure1, memory, true);
+                Marshal.StructureToPtr(structure, memory, false);
+                Marshal.StructureToPtr(structure, memory, true);
             }
             finally
             {
                 Marshal.FreeHGlobal(memory);
```
Comment (luqunl) [inline src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/StructureToPtrTests.cs]: forgot to call DestroyStructure to release its field memory?

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/PtrToStructureTests.cs [hunk 2]
```diff
@@ -2,33 +2,262 @@
...
+            {
+                Marshal.StructureToPtr(structure, ptr, false);
+
+                SomeTestStruct result = Assert.IsType<SomeTestStruct>(Marshal.PtrToStructure(ptr, typeof(SomeTestStruct)));
+                Assert.Equal(10, result.i);
+                Assert.Equal("hello", result.s);
+            }
+            finally
+            {
+                Marshal.FreeHGlobal(ptr);
```
Comment (luqunl) [inline src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/PtrToStructureTests.cs]: please also free structure.s memory

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/IsComObjectTests.cs [hunk 3]
```diff
@@ -12,13 +14,33 @@ public partial class IsComObjectTests
...
+            MethodInfo method = typeof(IsComObjectTests).GetMethod(nameof(NonGenericMethod));
+            Delegate d = method.CreateDelegate(typeof(NonGenericDelegate));
+            yield return new object[] { d };
+
+            yield return new object[] { new KeyValuePair<string, int>("key", 10) };
+
+            AssemblyBuilder assemblyBuilder = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("Assembly"), AssemblyBuilderAccess.RunAndCollect);
+            ModuleBuilder moduleBuilder = assemblyBuilder.DefineDynamicModule("Module");
+            TypeBuilder typeBuilder = moduleBuilder.DefineType("Type");
+            Type collectibleType = typeBuilder.CreateType();
```
Comment (luqunl) [inline src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/IsComObjectTests.cs]: how about create a scenario where type is dynamic created and it contains [ComImport] attribute?

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetTypedObjectForIUnknownTests.cs [hunk 4]
```diff
@@ -73,38 +124,86 @@ public static IEnumerable<object[]> GetTypedObjectForIUnknown_Invalid_TestData()
...
+
         [Theory]
+        [MemberData(nameof(GetTypedObjectForIUnknown_ArrayObjects_TestData))]
         [PlatformSpecific(TestPlatforms.Windows)]
-        [InlineData(typeof(AbstractClass))]
-        [InlineData(typeof(NonGenericClass))]
-        [InlineData(typeof(NonGenericStruct))]
-        [InlineData(typeof(NonGenericInterface))]
-        public void GetTypedObjectForIUnknown_UncastableType_ThrowsInvalidCastException(Type type)
+        public void GetTypedObjectForIUnknown_ArrayType_ThrowsBadImageFormatException(object o)
```
Comment (luqunl) [inline src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetTypedObjectForIUnknownTests.cs]: Could you create a new issue for this case?

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/PtrToStructureTests.cs [hunk 5]
```diff
@@ -2,33 +2,262 @@
...
+                Marshal.StructureToPtr(structure, ptr, false);
+
+                var result = new SequentialClass();
+                Marshal.PtrToStructure(ptr, result);
+                Assert.Equal(10, result.i);
+                Assert.Equal("hello", result.s);
+            }
+            finally
+            {
+                Marshal.FreeHGlobal(ptr);
```
Comment (luqunl) [inline src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/PtrToStructureTests.cs]: please release struct.s memory.

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetNativeVariantForObjectTests.cs [hunk 6]
```diff
@@ -103,55 +356,170 @@ public static void CharTypeNegative()
...
+            {
+                Assert.Throws<NotSupportedException>(() => Marshal.GetNativeVariantForObject(o, pNative));
+            }
+            finally
+            {
+                Marshal.FreeHGlobal(pNative);
+            }
+        }
+
+        public struct Variant
```
Comment (luqunl) [inline src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetNativeVariantForObjectTests.cs]: suggest move this type into common\

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetNativeVariantForObjectTests.cs [hunk 7]
```diff
@@ -2,98 +2,351 @@
...
+        {
+            var v = new Variant();
+            IntPtr pNative = Marshal.AllocHGlobal(Marshal.SizeOf(v));
+            try
+            {
+                Marshal.GetNativeVariantForObject(obj, pNative);
+
+                Variant result = Marshal.PtrToStructure<Variant>(pNative);
+                Assert.Equal(VarEnum.VT_BSTR, (VarEnum)result.vt);
+                Assert.Equal(obj, Marshal.PtrToStringBSTR(result.bstrVal));
```
Comment (luqunl) [inline src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetNativeVariantForObjectTests.cs]: Free memory

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetTypedObjectForIUnknownTests.cs (hunk 4 above)
Comment (hughbe) [reply src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetTypedObjectForIUnknownTests.cs]: What do you mean? The test passes on both .NET Framework and .NET Core. Do you mean making an issue for the fact it throws a BIFE?

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/StructureToPtrTests.cs (hunk 1 above)
Comment (hughbe) [reply src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/StructureToPtrTests.cs]: Fixed

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/PtrToStructureTests.cs (hunk 5 above)
Comment (hughbe) [reply src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/PtrToStructureTests.cs]: Fixed

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetNativeVariantForObjectTests.cs (hunk 6 above)
Comment (hughbe) [reply src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetNativeVariantForObjectTests.cs]: Done

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetNativeVariantForObjectTests.cs (hunk 7 above)
Comment (hughbe) [reply src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetNativeVariantForObjectTests.cs]: Fixed

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/IsComObjectTests.cs (hunk 3 above)
Comment (hughbe) [reply src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/IsComObjectTests.cs]: Sure - done

**Comment by karelz** [comment]
Welcome back @hughbe and thanks for pushing it further! :)

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetTypedObjectForIUnknownTests.cs (hunk 4 above)
Comment (luqunl) [reply src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetTypedObjectForIUnknownTests.cs]: >>Do you mean making an issue for the fact it throws a BIFE? 

Yes, please

**Pull request merged**